File: packages/core/src/core/feedback/fopdt.py
import math

import numpy as np
from sleipnir.autodiff import log, exp
from sleipnir.optimization import ExitStatus, Problem

from core.sample.logged_sample import LoggedSample


def fopdt_dynamics(t, K, tau, theta):
    # Softplus provides a continuously differentiable approximation for max(0.0, t - theta)
    t_diff = t - theta
    print(type(t_diff))
    t_eff = log(1.0 + exp(20.0 * t_diff)) / 20.0

    smooth_step = 1.0 / (1.0 + exp(-50.0 * t_diff))
    return K * (1.0 - exp(-t_eff / tau)) * smooth_step


class FOPDTSolver:
    def __init__(self, data: list[LoggedSample], dt: float):
        self.dt = dt
        self.N = len(data)
        self.u = np.array([sample.voltage for sample in data])
        self.y_meas = np.array([sample.velocity for sample in data])
        self.t_vec = np.arange(self.N) * self.dt

    def fit(self):
        problem = Problem()

        K = problem.decision_variable()
        tau = problem.decision_variable()
        theta = problem.decision_variable()

        K.set_value(self.y_meas[-1])
        tau.set_value(1.0)
        theta.set_value(0.1)

        max_time = float(self.N * self.dt)
        problem.subject_to(K >= 0.0)
        problem.subject_to(K <= 100.0)
        problem.subject_to(tau >= 1e-4)
        problem.subject_to(tau <= max_time)
        problem.subject_to(theta >= 0.0)
        problem.subject_to(theta <= max_time)

        cost = 0.0
        for t, y_m in zip(self.t_vec, self.y_meas):
            y_pred = fopdt_dynamics(t, K, tau, theta)
            cost += (y_m - y_pred) ** 2

        problem.minimize(cost)

        status = problem.solve()

        if status == ExitStatus.SUCCESS:
            K_fit = K.value()
            tau_fit = tau.value()
            theta_fit = theta.value()

            print("Optimization converged successfully.")
            print(f"Fitted Gain (K):          {K_fit:.4f}")
            print(f"Fitted Time Constant (τ): {tau_fit:.4f} s")
            print(f"Fitted Dead Time (θ):     {theta_fit:.4f} s")
            print(f"Final Mean Squared Error: {cost.value() / self.N:.6f}")
            return result_vals(K_fit, tau_fit, theta_fit, status)
        else:
            print(f"Solver failed to converge: {status.exit_status}")
            return None


def result_vals(K, tau, theta, status):
    class Result:
        pass
    res = Result()
    res.x = np.array([K, tau, theta])
    res.success = True
    res.status = status
    return res

File: src/core/feedback/fopdt.cpp
//
// Created by adarw on 7/21/26.
//

#include "fopdt.hpp"

#include <iostream>

namespace sysid {

    template <NumericCompatible T>
    T fopdt_dynamics(T t, T K, T tau, T theta) {
        T t_diff = t - theta;
        T t_eff = slp::log(1.0 + slp::exp(20.0 * t_diff)) / 20.0;

        T smooth_step = 1.0 / (1.0 + slp::exp(-50.0 * t_diff));
        return K * (1.0 - slp::exp(-t_eff / tau)) * smooth_step;
    }


    FOPDTSolver::FOPDTSolver(const System& system) : system(system) {
        if (system.systemType == SystemType::POSITIONAL) {
            std::cerr << "Warning: A FOPDT model shouldn't be fitted to a positional system." << std::endl;
        }
        K = problem.decision_variable();
        tau = problem.decision_variable();
        theta = problem.decision_variable();
    }

    void FOPDTSolver::makeCostFunction(SampleVectorPtr log) {
        J = 0;
        for (auto [idx, y_meas] : std::views::enumerate(log->y_meas)) {
            const dvar t = idx * system.dt;
            dvar y_pred = fopdt_dynamics(t, K, tau, theta);
            J += slp::pow(y_meas - y_pred, 2);
        }

        problem.minimize(J);
    }

    void FOPDTSolver::putData(SampleVectorPtr log) {
        K.set_value(log->y_meas(log->N-1)); // Initial guess for K
        tau.set_value(system.dt * log->N / 2); // Initial guess for tau
        theta.set_value(5*system.dt); // Initial guess for theta

        double max_time = log->N * system.dt;
        problem.subject_to(K >= 0);
        problem.subject_to(K <= log->y_meas.maxCoeff());

        problem.subject_to(tau >= 1e-4);
        problem.subject_to(tau <= max_time);

        problem.subject_to(theta >= 0);
        problem.subject_to(theta <= max_time);

        makeCostFunction(log);
    }

    OptimizationResult<OLSMetrics, FOPDTGains> FOPDTSolver::solve() {
        if (const slp::ExitStatus status = problem.solve(); status != slp::ExitStatus::SUCCESS) {
            throw std::runtime_error("FOPDT solver could not converge to a solution!");
        }
        return OptimizationResult<OLSMetrics, FOPDTGains> {
            {
                0,
                J.value()
            },
            {
                K.value(),
                tau.value(),
                theta.value()
            }
        };
    }
} // sysid